File: frontend/src/types/site.ts
import type { PhpVersion } from './phpVersion'
import type { SiteLogStreamHandlers, TailSiteLogOptions } from './siteLog'

/**
 * What serves a site's content, mirroring the backend's
 * `Maran.Modules.Sites.Domain.Enums.SiteBackendType`. The API serializes enums as camelCase
 * strings (panel-wide `JsonStringEnumConverter`), so these are the camelCase forms of the C#
 * member names.
 *
 * There is deliberately no "unspecified" member: a site whose backend cannot be named is a site
 * whose vhost cannot be rendered.
 */
export type SiteBackendType = 'static' | 'php' | 'reverseProxy'

/**
 * Whether a site serves its own content, mirroring the backend's
 * `Maran.Modules.Sites.Domain.Enums.SiteStatus`.
 *
 * A serving state, not a lifecycle state: a disabled site keeps its vhost, its aliases and its
 * log paths, and answers with a suspension response. A site that no longer exists has no row.
 */
export type SiteStatus = 'enabled' | 'disabled'

/**
 * Outward, list-shaped view of a site, mirroring the backend's `SiteDto` field-for-field.
 *
 * Without the document root, on purpose: the backend keeps a filesystem path out of its
 * widest-read surface, and {@link SiteDetail} carries it for the single-site read.
 */
export interface Site {
  /** The site's identity. */
  id: string
  /** The account that owns this site. */
  accountId: string
  /** The primary domain served by this site. */
  domain: string
  /** Which backend serves this site's content. */
  backendType: SiteBackendType
  /** The bound PHP version, or the empty string when the backend is not PHP. */
  phpVersion: string
  /** Whether the site serves its own content or a suspension response. */
  status: SiteStatus
  /** The instant the site was created, as an ISO-8601 string. */
  createdAt: string
}

/**
 * Single-site view, mirroring the backend's `SiteDetailDto` field-for-field. It is a superset of
 * {@link Site}, but is not declared as one: the two are separate contracts on the backend, and
 * tying them here would make an additive change to one silently change the other.
 */
export interface SiteDetail {
  /** The site's identity. */
  id: string
  /** The account that owns this site. */
  accountId: string
  /** The primary domain served by this site. */
  domain: string
  /** Additional hostnames answered by the same vhost. */
  aliases: string[]
  /** Which backend serves this site's content. */
  backendType: SiteBackendType
  /** The bound PHP version, or the empty string when the backend is not PHP. */
  phpVersion: string
  /** The upstream forwarded to, or the empty string when the backend is not a reverse proxy. */
  proxyUpstream: string
  /** The absolute document root the agent allocated. */
  documentRoot: string
  /** Whether a TLS certificate is currently installed for this site. */
  hasCertificate: boolean
  /** Whether the site serves its own content or a suspension response. */
  status: SiteStatus
  /** The instant the site was created, as an ISO-8601 string. */
  createdAt: string
}

/**
 * Request body for `POST /api/v1/sites`, mirroring the backend's `CreateSiteRequest`
 * field-for-field. The caller's address and user agent are NOT here: the backend reads those
 * from the connection, so the request being audited cannot set them.
 */
export interface CreateSiteRequest {
  /** The account that will own the site. */
  accountId: string
  /** The primary domain the site serves. */
  domain: string
  /** Additional hostnames answered by the same vhost. */
  aliases: string[]
  /** Which backend serves the site's content. */
  backendType: SiteBackendType
  /** The installed PHP version to bind to; required when the backend is PHP. */
  phpVersion: string
  /** The upstream to forward to; required when the backend is a reverse proxy. */
  proxyUpstream: string
}

/**
 * Request body for `POST /api/v1/sites/{id}/php-version`, mirroring the backend's
 * `ChangeSitePhpVersionRequest`.
 */
export interface ChangeSitePhpVersionRequest {
  /** The installed version to switch to. */
  phpVersion: string
}

/**
 * Typed access to the sites endpoints.
 *
 * Called from Pinia stores only — never from a component (rules/vue.md).
 */
export interface SitesApi {
  /**
   * Lists the sites the caller may see.
   * @param signal Optional abort signal to cancel the in-flight request.
   * @returns The sites, in the order the panel reports them.
   */
  list: (signal?: AbortSignal) => Promise<Site[]>

  /**
   * Reads one site. Another customer's site answers 404, never 403.
   * @param id The site's identity.
   * @param signal Optional abort signal to cancel the in-flight request.
   * @returns The site's full detail.
   */
  get: (id: string, signal?: AbortSignal) => Promise<SiteDetail>

  /**
   * Creates a site: its document root, vhost and pool on the host, then the row.
   * @param request The site's account, domain, aliases and backend.
   * @param signal Optional abort signal to cancel the in-flight request.
   * @returns The site as the panel created it.
   */
  create: (request: CreateSiteRequest, signal?: AbortSignal) => Promise<Site>

  /**
   * Rebinds a site to a different installed PHP version.
   * @param id The site's identity.
   * @param phpVersion The installed version to switch to.
   * @param signal Optional abort signal to cancel the in-flight request.
   * @returns The site in its new state.
   */
  changePhpVersion: (id: string, phpVersion: string, signal?: AbortSignal) => Promise<Site>

  /**
   * Returns a disabled site to serving its own content. Idempotent.
   * @param id The site's identity.
   * @param signal Optional abort signal to cancel the in-flight request.
   * @returns The site in its new state.
   */
  enable: (id: string, signal?: AbortSignal) => Promise<Site>

  /**
   * Makes a site serve a suspension response instead of its content. Idempotent.
   * @param id The site's identity.
   * @param signal Optional abort signal to cancel the in-flight request.
   * @returns The site in its new state.
   */
  disable: (id: string, signal?: AbortSignal) => Promise<Site>

  /**
   * Removes a site's vhost and its row. The customer's files are left alone.
   * @param id The site's identity.
   * @param signal Optional abort signal to cancel the in-flight request.
   * @returns Whether the panel removed the site.
   */
  remove: (id: string, signal?: AbortSignal) => Promise<boolean>

  /**
   * Lists the PHP versions installed on this server — the reference data the site form selects
   * from, so nobody types a version the host does not have.
   * @param signal Optional abort signal to cancel the in-flight request.
   * @returns The installed versions.
   */
  listPhpVersions: (signal?: AbortSignal) => Promise<PhpVersion[]>

  /**
   * Tails one of a site's logs: recent lines first, then new ones as they are written.
   *
   * Resolves only once the stream has ended and `onEnd` has been called, so a caller can await
   * teardown; aborting the signal ends the stream with `cancelled` rather than throwing.
   * @param options The site, the log and how much history to replay.
   * @param handlers Where lines and the ending are delivered.
   * @param signal Abort signal that stops the stream and releases the connection.
   * @returns Resolves once the stream has ended.
   */
  tailLog: (
    options: TailSiteLogOptions,
    handlers: SiteLogStreamHandlers,
    signal: AbortSignal,
  ) => Promise<void>
}

File: backend/src/Maran.Modules/Sites/Domain/Entities/Site.cs
using Maran.Modules.Sites.Domain.Enums;

namespace Maran.Modules.Sites.Domain.Entities;

/// <summary>
/// A website served from one account's home directory: a domain, the aliases answering with it, the
/// backend that renders it, and whether it is currently serving (spec §11).
/// </summary>
/// <remarks>
/// The agent keeps no record of any of this — the vhost on disk is a RENDERING of this row, not a
/// second copy of it — so this entity is the site's only definition, and every agent call that
/// re-renders a vhost is handed these facts. That is also why nothing here has a public setter: a
/// field assigned from outside is a field that can disagree with what was rendered.
/// </remarks>
public sealed class Site
{
    /// <summary>The hostnames this site claims across the server: its domain and every alias.</summary>
    /// <remarks>
    /// The backing field EF Core reads and writes, and the only place the claims are built. Kept
    /// private so that no caller can claim a name without a site or leave a site without its
    /// claims: <see cref="SiteHostname"/> explains why an unclaimed alias is a domain takeover.
    /// </remarks>
    private readonly List<SiteHostname> _hostnames = [];

    /// <summary>The site's identity.</summary>
    public Guid Id { get; private set; }

    /// <summary>The account that owns this site. Every tenant-scoped query is closed over this column.</summary>
    public Guid AccountId { get; private set; }

    /// <summary>The primary domain served by this site, unique across the server.</summary>
    public string Domain { get; private set; }

    /// <summary>Additional hostnames answered by the same vhost.</summary>
    public IReadOnlyList<string> Aliases { get; private set; }

    /// <summary>
    /// Every hostname this site claims — <see cref="Domain"/> and each of <see cref="Aliases"/> —
    /// as rows a unique key makes exclusive across the whole server.
    /// </summary>
    /// <remarks>
    /// Not a second copy of the two properties above for a reader's convenience: those are what the
    /// vhost is RENDERED from, and this is what a name being taken is DECIDED from, which has to be
    /// a key in the database because a check in a handler cannot be atomic with the insert that
    /// follows it (<see cref="SiteHostname"/>).
    /// </remarks>
    public IReadOnlyCollection<SiteHostname> Hostnames
    {
        get
        {
            return _hostnames;
        }
    }

    /// <summary>Which backend serves this site's content.</summary>
    public SiteBackendType BackendType { get; private set; }

    /// <summary>The bound PHP version, or the empty string when the backend is not PHP.</summary>
    public string PhpVersion { get; private set; }

    /// <summary>The upstream forwarded to, or the empty string when the backend is not a reverse proxy.</summary>
    public string ProxyUpstream { get; private set; }

    /// <summary>The absolute document root the agent allocated. Operator-facing; never rendered into a customer's error.</summary>
    public string DocumentRoot { get; private set; }

    /// <summary>
    /// Whether a TLS certificate is currently installed for this site.
    /// </summary>
    /// <remarks>
    /// NOT dead code, and not to be removed for being always false today: it is the fact every
    /// vhost re-render must be told, because a renderer that guesses drops a live site back to
    /// plain HTTP on the next unrelated edit. Nothing in the Sites module writes it — the Ssl
    /// module owns that write path and flips it when a certificate is installed or removed. Until
    /// then it is honestly false, which is the only safe answer while no certificate exists.
    /// </remarks>
    public bool HasCertificate { get; private set; }

    /// <summary>Whether the site serves its own content or a suspension response.</summary>
    public SiteStatus Status { get; private set; }

    /// <summary>The instant the site was created.</summary>
    public DateTimeOffset CreatedAt { get; private set; }

    /// <summary>Creates a site in the <see cref="SiteStatus.Enabled"/> state with no certificate.</summary>
    /// <param name="id">The site's identity.</param>
    /// <param name="accountId">The account that owns this site.</param>
    /// <param name="domain">The primary domain served by this site.</param>
    /// <param name="aliases">Additional hostnames answered by the same vhost.</param>
    /// <param name="backendType">Which backend serves this site's content.</param>
    /// <param name="phpVersion">The bound PHP version, or the empty string when the backend is not PHP.</param>
    /// <param name="proxyUpstream">The upstream forwarded to, or the empty string when the backend is not a proxy.</param>
    /// <param name="documentRoot">The absolute document root the agent allocated.</param>
    /// <param name="createdAt">The instant the site was created, taken from <see cref="IClock"/>.</param>
    public Site(
        Guid id,
        Guid accountId,
        string domain,
        IReadOnlyList<string> aliases,
        SiteBackendType backendType,
        string phpVersion,
        string proxyUpstream,
        string documentRoot,
        DateTimeOffset createdAt)
    {
        Id = id;
        AccountId = accountId;
        Domain = domain;
        Aliases = [.. aliases];
        BackendType = backendType;
        PhpVersion = phpVersion;
        ProxyUpstream = proxyUpstream;
        DocumentRoot = documentRoot;
        HasCertificate = false;
        Status = SiteStatus.Enabled;
        CreatedAt = createdAt;

        // Built here and nowhere else: a site that exists has claimed every name it answers for,
        // by construction rather than by a handler remembering to write a second row.
        _hostnames.Add(new SiteHostname(domain, this));
        foreach (var alias in Aliases)
        {
            _hostnames.Add(new SiteHostname(alias, this));
        }
    }

    /// <summary>Parameterless constructor required by EF Core materialization.</summary>
    private Site()
    {
        Domain = string.Empty;
        Aliases = [];
        PhpVersion = string.Empty;
        ProxyUpstream = string.Empty;
        DocumentRoot = string.Empty;
    }

    /// <summary>Rebinds the site to a different installed PHP version.</summary>
    /// <param name="version">The installed version to bind to.</param>
    public void ChangePhpVersion(string version)
    {
        PhpVersion = version;
    }

    /// <summary>Returns the site to serving its own content. Idempotent: enabling an enabled site is a no-op.</summary>
    public void Enable()
    {
        Status = SiteStatus.Enabled;
    }

    /// <summary>Makes the site serve a suspension response. Idempotent: disabling a disabled site is a no-op.</summary>
    public void Disable()
    {
        Status = SiteStatus.Disabled;
    }

    /// <summary>Records that a TLS certificate is now installed for this site.</summary>
    /// <remarks>
    /// The write path <see cref="HasCertificate"/> describes. It exists here, on the entity, because
    /// the Ssl module (which owns certificates) may not reach into this module's rows any other way,
    /// and because a field with no way to become true is a field no test can hold the renderer to.
    /// Nothing in the Sites module calls it; Task 14's certificate installation does.
    /// </remarks>
    public void AttachCertificate()
    {
        HasCertificate = true;
    }

    /// <summary>Records that this site no longer has a TLS certificate.</summary>
    /// <remarks>The other half of <see cref="AttachCertificate"/>; removal must be expressible too, or a
    /// revoked certificate would leave the vhost rendering a TLS block for a file that is gone.</remarks>
    public void DetachCertificate()
    {
        HasCertificate = false;
    }
}
